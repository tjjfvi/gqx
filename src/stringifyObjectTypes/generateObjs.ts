
import { Context, Obj, Id, Prop } from ".";
import unwrapType from "../unwrapType";

export default (ctx: Context): [Obj[], Id[], Context] => {
  const ids: Id[] = [];
  const objs: Obj[] =
    Object.keys(ctx.objectTypes)
      .sort()
      .map(name => {
        const { fields, unions, implements: impls } = ctx.objectTypes[name];
        const props = fields.map(field => {
          const id: Id = {
            type: name,
            prop: field.name.value,
          };
          ids.push(id);
          const [{ name: { value: typeName } }, wrap,, wrapHKT] = unwrapType(field.type);
          const shallow = !(typeName in ctx.objectTypes);
          let prop: Prop = {
            id,
            type: typeName,
            shallow,
            wrap,
            wrapHKT,
            args: field.arguments?.slice() ?? [],
            directives: field.directives?.slice() ?? [],
            loc: field.loc,
          };
          return prop;
        }).sort((p, q) => p.id.prop > q.id.prop ? 1 : -1);
        const shallowProps = props.filter(p => p.shallow);
        const deepProps = props.filter(p => !p.shallow);
        const obj: Obj = {
          type: name,
          shallowProps,
          deepProps,
          props,
          unions,
          implements: impls,
          isBase: ctx.baseTypes.includes(name),
        };
        return obj;
      });
  return [objs, ids, ctx];
}
