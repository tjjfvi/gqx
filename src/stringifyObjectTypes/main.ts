
import { unwrapType, Context, stringifyFragResult, stringifyIds, stringifyFragEnums } from "."

interface Id {
  type: string;
  prop: string;
}

interface Prop {
  id: Id;
  type: string;
  shallow: boolean;
  wrap: (x: string) => string;
}

interface Obj {
  type: string;
  shallowProps: Prop[];
  deepProps: Prop[];
}

export { Id, Prop, Obj };

export default (ctx: Context) => {
  const ids: Id[] = [];

  const objs: Obj[] =
    Object.keys(ctx.objectTypes)
      .filter(k => !ctx.baseTypes.includes(k))
      .sort()
      .map(name => {
        const fields = ctx.objectTypes[name];
        const props = fields.map(field => {
          const id: Id = {
            type: name,
            prop: field.name.value,
          };
          ids.push(id);
          const [{ name: { value: typeName } }, wrap] = unwrapType(field.type);
          const shallow = typeName in ctx.objectTypes;
          let prop: Prop = {
            id,
            type: typeName,
            shallow,
            wrap,
          };
          return prop;
        }).sort((p, q) => p.id.prop > q.id.prop ? 1 : -1);
        const shallowProps = props.filter(p => p.shallow);
        const deepProps = props.filter(p => !p.shallow);
        const obj: Obj = {
          type: name,
          shallowProps,
          deepProps,
        };
        return obj;
      })

  return [
    stringifyIds(ids),
    stringifyFragEnums(objs, ctx),
    ...objs.map(o => stringifyFragResult(o, ctx))
  ].join("\n\n");
}

