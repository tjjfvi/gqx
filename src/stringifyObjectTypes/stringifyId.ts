
import { Id } from ".";

export default (id: Id) => `"${id.type}$${id.prop}"`
