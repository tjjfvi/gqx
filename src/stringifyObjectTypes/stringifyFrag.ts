/* eslint-disable @typescript-eslint/no-non-null-assertion */

import { Obj } from ".";
import { stringifyId } from "./stringifyId";
import { indent } from "../indent";

export const stringifyFrag = (obj: Obj, objs: Obj[]) => `
interface _$${obj.type}$ extends $Hkt {
  return: ${obj.type}.$<this["input"]>
}

interface _$${obj.type}<T extends $Hkt> {
  $<F extends ${obj.type}.$>(frag: F): $CallHkt<T, ${obj.type}<F>>
  $<F extends ${obj.type}.$[]>(...frag: F): $CallHkt<T, ${obj.type}<F>>
${
  obj.props.map(prop =>
    indent`${prop.id.prop}: ${
      prop.shallow ?
        `$CallHkt<T, $Wrap<$ShallowPropTypes[${stringifyId(prop.id)}]>[${stringifyId(prop.id)}]>` :
        `_$${prop.type}<$Compose<T, $WrapHkt<${stringifyId(prop.id)}>>>`
    }`
  ).join("\n")
}
}

const _$${obj.type} = <T extends $Hkt>(f: (x: $Fragment) => any): _$${obj.type}<T> => ({
  $: ((...a: $Fragment[]) => f($Fragment.merge(...a))) as any,
${
  obj.props.map(prop =>
    indent`get ${prop.id.prop}(){ return ${
      prop.shallow ?
        `f($Fragment._create(${stringifyId(prop.id)}, null))` :
        `_$${prop.type}<$Compose<T, $WrapHkt<${stringifyId(prop.id)}>>>(x => $Fragment._create(${stringifyId(prop.id)}, f(x)))`
    } },`
  ).join("\n")
}
})

export const ${obj.type} = _$${obj.type}<_$${obj.type}$>(x => x)

// @ts-ignore${/* eslint-disable-next-line max-len */""}
export type ${obj.type}<F extends ${obj.type}.$ | ${obj.type}.$[]> = F extends ${obj.type}.$<infer X> ? X : $ExpandDeep<{ [K in keyof F]: (x: ${obj.type}<F[K]>) => void }[number] extends (x: infer X) => void ? X : never>

export declare namespace ${obj.type} {
  export interface $<out T = unknown> extends $Fragment {
    [$fragTypes]: {
      _${obj.type}: T
${
  obj.supertypes.map(type =>
    indent(indent(indent`_${type}: T${
      objs
        .find(x => x.type === type)!
        .subtypes
        .filter(x => x !== obj.type)
        .map(x => ` | { __typename: ${JSON.stringify(x)} }`)
        .join("")
    }`))
  )
}
    }
  }
}
`.trim()
