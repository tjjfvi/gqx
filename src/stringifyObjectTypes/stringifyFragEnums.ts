

import { Obj, Context } from ".";
import stringifyId from "./stringifyId";

export default (objs: Obj[], ctx: Context) => {
  ctx.exports.push("$_");
  return [
    ...objs.map(obj => {
      ctx.exports.push(obj.type + "$");
      return [
        `type ${obj.type}$ =\n`,
        ...obj.shallowProps.map(p => `  | typeof ${stringifyId(p.id)}\n`),
        ...obj.deepProps.map(p => `  | $<typeof ${stringifyId(p.id)}, ${p.type}$>\n`),
      ].join("");
    }),
    `\ntype $_ =${objs.map(o => `\n  | ${o.type}$`).join("")}`,
  ].join("\n")
}
