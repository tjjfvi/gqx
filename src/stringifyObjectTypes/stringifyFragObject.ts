
import { Obj, Prop } from ".";
import stringifyId from "./stringifyId";
import indent from "../indent";

export default (obj: Obj) => {
  const typeofType = `\n\ntype typeof$${obj.type} = typeof ${obj.type}`

  const fragCast = `const ${obj.type}$ = <F extends $$DeepArray<${obj.type}$>>(...frag: F) => [...frag];\n\n`

  const shallowObj = genObj(
    obj.shallowProps,
    p => stringifyId(p.id),
    indent`$: ${obj.type}$,\n`
  ) + " as const";

  const deepType = genObj(obj.deepProps, p =>
    `$$MapWrap<typeof$${p.type}, ${stringifyId(p.id)}>`,
  obj.shallowProps.length ? "" : indent`$: typeof ${obj.type}$,\n`);

  const deepObj = genObj(obj.deepProps, p =>
    `$$mapWrap(() => ${p.type}, ${stringifyId(p.id)})`,
  obj.shallowProps.length ? indent`..._${obj.type},\n` : indent`$: ${obj.type}$,\n`);

  if(!obj.deepProps.length)
    return fragCast + `export const ${obj.type} = ${shallowObj};` + typeofType;

  if(!obj.shallowProps.length)
    return fragCast + `export const ${obj.type}: ${deepType} = ${deepObj};` + typeofType;


  return fragCast + `
const _${obj.type} = ${shallowObj};

export const ${obj.type}: typeof _${obj.type} & ${deepType} = ${deepObj};
`.trim() + typeofType;
}

function genObj(props: Prop[], f: (p: Prop) => string, str = ""){
  if(!props.length)
    return "{}";
  return `{\n${str}${
    props.map(p =>
      indent`${p.id.prop}: ${f(p)},`
    ).join("\n")
  }\n}`;
}
