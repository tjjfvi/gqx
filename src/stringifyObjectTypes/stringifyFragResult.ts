
import { Context, Obj, Prop } from ".";
import stringifyId from "./stringifyId";

type P = [Prop, string, string];

export default (obj: Obj, ctx: Context) => {
  ctx.exports.push("$" + obj.type);
  const shallows = obj.shallowProps.map((prop): P =>
    [prop, "typeof " + stringifyId(prop.id), prop.type]
  );
  const deeps = obj.deepProps.map((prop): [Prop, (x: string) => string] =>
    [prop, x => `$<typeof ${stringifyId(prop.id)}, ${x}>`]
  ).map(([prop, sub]): P =>
    [prop, sub("any"), `$$${prop.type}<Extract<F, ${sub(prop.type + "$")}>["l"]>`]
  );
  const core = `type $$${obj.type}<F extends ${obj.type}$> =\n${[...shallows, ...deeps].map(x =>
    `  & (${x[1]} extends F ? { ${x[0].id.prop}: ${x[0].wrap(x[2])} } : {})\n`
  ).join("")}`;
  const alias = `type $${obj.type}<F extends ${obj.type}$[]> = $$${obj.type}<F[number]>;`;
  return core + alias;
}
