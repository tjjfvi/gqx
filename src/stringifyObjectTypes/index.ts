
import { Context } from "..";
import stringifyFragEnums from "./stringifyFragEnums";
import stringifyFragResult from "./stringifyFragResult";
import { stringifyObjectTypeInfo } from "./stringifyObjectTypeInfo";
import stringifyFragObject from "./stringifyFragObject";
import generateObjs from "./generateObjs";

interface Id {
  type: string;
  prop: string;
}

interface Prop {
  id: Id;
  type: string;
  shallow: boolean;
  wrap: (x: string) => string;
  wrapHKT: string;
}

interface Obj {
  type: string;
  shallowProps: Prop[];
  deepProps: Prop[];
}

export { Id, Prop, Obj, Context };

export default (ctx: Context) => {
  const [objs] = generateObjs(ctx);

  return [
    stringifyObjectTypeInfo(objs),
    stringifyFragEnums(objs),
    ...objs.map(o => stringifyFragObject(o)),
    ...objs.map(o => stringifyFragResult(o))
  ].join("\n\n");
}

