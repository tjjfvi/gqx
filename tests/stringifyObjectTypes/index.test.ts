
import stringifyObjectTypes from "../../src/stringifyObjectTypes";

test("", () => {
  const ctx = utils.getOperations(["stringifyObjectTypes"]);
  const code = stringifyObjectTypes(ctx);
  expect(code).toParse().toMatchSnapshot();
})
