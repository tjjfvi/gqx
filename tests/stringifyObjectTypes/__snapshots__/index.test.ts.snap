// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `
"export type $$ObjectType = (
  | \\"BinaryTree\\"
  | \\"DeepOnly\\"
  | \\"Query\\"
  | \\"Recursive\\"
  | \\"ShallowOnly\\"
)

export const $$objectTypeInfoMap = {
  BinaryTree: new $$ObjectTypeInfo(
    {
      BinaryTree$node: \\"node\\",
    },
    {
      node: \\"BinaryTree$node\\",
    },
    null as any as {
      BinaryTree$node: \\"String\\",
    },
    null as any as {
      BinaryTree$node: $$Identity,
    },
    {
      BinaryTree$left: \\"left\\",
      BinaryTree$right: \\"right\\",
    },
    {
      left: \\"BinaryTree$left\\",
      right: \\"BinaryTree$right\\",
    },
    null as any as {
      BinaryTree$left: \\"BinaryTree\\",
      BinaryTree$right: \\"BinaryTree\\",
    },
    null as any as {
      BinaryTree$left: $$Optional<$$Identity>,
      BinaryTree$right: $$Optional<$$Identity>,
    },
    null as any as {
      BinaryTree$left: {},
      BinaryTree$node: {},
      BinaryTree$right: {},
    },
    {
      BinaryTree$left: {},
      BinaryTree$node: {},
      BinaryTree$right: {},
    },
    {
      BinaryTree$left: {},
      BinaryTree$node: {},
      BinaryTree$right: {},
    },
    {
      BinaryTree$left: \\"GraphQL request\\",
      BinaryTree$node: \\"GraphQL request\\",
      BinaryTree$right: \\"GraphQL request\\",
    } as const,
  ),
  DeepOnly: new $$ObjectTypeInfo(
    {}, {}, {}, {},
    {
      DeepOnly$a: \\"a\\",
      DeepOnly$b: \\"b\\",
    },
    {
      a: \\"DeepOnly$a\\",
      b: \\"DeepOnly$b\\",
    },
    null as any as {
      DeepOnly$a: \\"ShallowOnly\\",
      DeepOnly$b: \\"ShallowOnly\\",
    },
    null as any as {
      DeepOnly$a: $$Identity,
      DeepOnly$b: $$Array<$$Optional<$$Identity>>,
    },
    null as any as {
      DeepOnly$a: {},
      DeepOnly$b: {},
    },
    {
      DeepOnly$a: {},
      DeepOnly$b: {},
    },
    {
      DeepOnly$a: {},
      DeepOnly$b: {},
    },
    {
      DeepOnly$a: \\"GraphQL request\\",
      DeepOnly$b: \\"GraphQL request\\",
    } as const,
  ),
  Query: new $$ObjectTypeInfo(
    {
      Query$oneDirective: \\"oneDirective\\",
    },
    {
      oneDirective: \\"Query$oneDirective\\",
    },
    null as any as {
      Query$oneDirective: \\"String\\",
    },
    null as any as {
      Query$oneDirective: $$Optional<$$Identity>,
    },
    {
      Query$multDirectivesArgs: \\"multDirectivesArgs\\",
      Query$multDirectivesNoArgs: \\"multDirectivesNoArgs\\",
      Query$multDirectivesNullArgs: \\"multDirectivesNullArgs\\",
      Query$noDirectives: \\"noDirectives\\",
    },
    {
      multDirectivesArgs: \\"Query$multDirectivesArgs\\",
      multDirectivesNoArgs: \\"Query$multDirectivesNoArgs\\",
      multDirectivesNullArgs: \\"Query$multDirectivesNullArgs\\",
      noDirectives: \\"Query$noDirectives\\",
    },
    null as any as {
      Query$multDirectivesArgs: \\"BinaryTree\\",
      Query$multDirectivesNoArgs: \\"ShallowOnly\\",
      Query$multDirectivesNullArgs: \\"DeepOnly\\",
      Query$noDirectives: \\"Recursive\\",
    },
    null as any as {
      Query$multDirectivesArgs: $$Identity,
      Query$multDirectivesNoArgs: $$Optional<$$Identity>,
      Query$multDirectivesNullArgs: $$Identity,
      Query$noDirectives: $$Array<$$Identity>,
    },
    null as any as {
      Query$multDirectivesArgs: {},
      Query$multDirectivesNoArgs: {},
      Query$multDirectivesNullArgs: {},
      Query$noDirectives: {},
      Query$oneDirective: {},
    },
    {
      Query$multDirectivesArgs: {},
      Query$multDirectivesNoArgs: {},
      Query$multDirectivesNullArgs: {},
      Query$noDirectives: {},
      Query$oneDirective: {},
    },
    {
      Query$multDirectivesArgs: {
        scalarArg: {
          a: \\"test\\",
          b: 12345,
        },
        inputArg: {
          a: {
            x: 5,
            y: [
              1,
              2,
              3,
            ],
            z: {
              x: 5,
            },
          },
        },
        scalarArrArg: {
          a: [
            \\"a\\",
            \\"b\\",
            \\"c\\",
            null,
          ],
        },
        inputArrArg: {
          a: [
            {
              x: 5,
            },
            {
              x: 6,
            },
            {
              y: [
                1,
                2,
                3,
              ],
            },
          ],
        },
      },
      Query$multDirectivesNoArgs: {
        noArgs: {},
        scalarArg: {},
        inputArg: {},
        scalarArrArg: {},
        inputArrArg: {},
      },
      Query$multDirectivesNullArgs: {
        scalarArg: {
          a: null,
          b: null,
        },
        inputArg: {
          a: null,
        },
        scalarArrArg: {
          a: null,
        },
        inputArrArg: {
          a: null,
        },
      },
      Query$noDirectives: {},
      Query$oneDirective: {
        scalarArg: {
          x: \\"string\\",
        },
      },
    },
    {
      Query$multDirectivesArgs: \\"GraphQL request\\",
      Query$multDirectivesNoArgs: \\"GraphQL request\\",
      Query$multDirectivesNullArgs: \\"GraphQL request\\",
      Query$noDirectives: \\"GraphQL request\\",
      Query$oneDirective: \\"GraphQL request\\",
    } as const,
  ),
  Recursive: new $$ObjectTypeInfo(
    {
      Recursive$id: \\"id\\",
    },
    {
      id: \\"Recursive$id\\",
    },
    null as any as {
      Recursive$id: \\"ID\\",
    },
    null as any as {
      Recursive$id: $$Identity,
    },
    {
      Recursive$next: \\"next\\",
    },
    {
      next: \\"Recursive$next\\",
    },
    null as any as {
      Recursive$next: \\"Recursive\\",
    },
    null as any as {
      Recursive$next: $$Identity,
    },
    null as any as {
      Recursive$id: {},
      Recursive$next: {},
    },
    {
      Recursive$id: {},
      Recursive$next: {},
    },
    {
      Recursive$id: {},
      Recursive$next: {},
    },
    {
      Recursive$id: \\"GraphQL request\\",
      Recursive$next: \\"GraphQL request\\",
    } as const,
  ),
  ShallowOnly: new $$ObjectTypeInfo(
    {
      ShallowOnly$a: \\"a\\",
      ShallowOnly$b: \\"b\\",
    },
    {
      a: \\"ShallowOnly$a\\",
      b: \\"ShallowOnly$b\\",
    },
    null as any as {
      ShallowOnly$a: \\"String\\",
      ShallowOnly$b: \\"Number\\",
    },
    null as any as {
      ShallowOnly$a: $$Identity,
      ShallowOnly$b: $$Identity,
    },
    {}, {}, {}, {},
    null as any as {
      ShallowOnly$a: {},
      ShallowOnly$b: {},
    },
    {
      ShallowOnly$a: {},
      ShallowOnly$b: {},
    },
    {
      ShallowOnly$a: {},
      ShallowOnly$b: {},
    },
    {
      ShallowOnly$a: \\"GraphQL request\\",
      ShallowOnly$b: \\"GraphQL request\\",
    } as const,
  ),
}

export type BinaryTree$ = $$ObjectTypeInfoMap[\\"BinaryTree\\"][\\"Frag\\"];
export type DeepOnly$ = $$ObjectTypeInfoMap[\\"DeepOnly\\"][\\"Frag\\"];
export type Query$ = $$ObjectTypeInfoMap[\\"Query\\"][\\"Frag\\"];
export type Recursive$ = $$ObjectTypeInfoMap[\\"Recursive\\"][\\"Frag\\"];
export type ShallowOnly$ = $$ObjectTypeInfoMap[\\"ShallowOnly\\"][\\"Frag\\"];

const BinaryTree$$ = <F extends $$DeepArray<BinaryTree$>>(...frag: F): F[number][] => frag;

const _BinaryTree = {
  $: BinaryTree$$,
  node: \\"BinaryTree$node\\",
} as const;

export const BinaryTree: typeof _BinaryTree & {
  left: $$MapWrap<typeof$BinaryTree, \\"BinaryTree$left\\">,
  right: $$MapWrap<typeof$BinaryTree, \\"BinaryTree$right\\">,
} = {
  ..._BinaryTree,
  left: $$mapWrap(() => BinaryTree, \\"BinaryTree$left\\"),
  right: $$mapWrap(() => BinaryTree, \\"BinaryTree$right\\"),
};

type typeof$BinaryTree = typeof BinaryTree

const DeepOnly$$ = <F extends $$DeepArray<DeepOnly$>>(...frag: F): F[number][] => frag;

export const DeepOnly: {
  $: typeof DeepOnly$$,
  a: $$MapWrap<typeof$ShallowOnly, \\"DeepOnly$a\\">,
  b: $$MapWrap<typeof$ShallowOnly, \\"DeepOnly$b\\">,
} = {
  $: DeepOnly$$,
  a: $$mapWrap(() => ShallowOnly, \\"DeepOnly$a\\"),
  b: $$mapWrap(() => ShallowOnly, \\"DeepOnly$b\\"),
};

type typeof$DeepOnly = typeof DeepOnly

const Query$$ = <F extends $$DeepArray<Query$>>(...frag: F): F[number][] => frag;

const _Query = {
  $: Query$$,
  oneDirective: \\"Query$oneDirective\\",
} as const;

export const Query: typeof _Query & {
  multDirectivesArgs: $$MapWrap<typeof$BinaryTree, \\"Query$multDirectivesArgs\\">,
  multDirectivesNoArgs: $$MapWrap<typeof$ShallowOnly, \\"Query$multDirectivesNoArgs\\">,
  multDirectivesNullArgs: $$MapWrap<typeof$DeepOnly, \\"Query$multDirectivesNullArgs\\">,
  noDirectives: $$MapWrap<typeof$Recursive, \\"Query$noDirectives\\">,
} = {
  ..._Query,
  multDirectivesArgs: $$mapWrap(() => BinaryTree, \\"Query$multDirectivesArgs\\"),
  multDirectivesNoArgs: $$mapWrap(() => ShallowOnly, \\"Query$multDirectivesNoArgs\\"),
  multDirectivesNullArgs: $$mapWrap(() => DeepOnly, \\"Query$multDirectivesNullArgs\\"),
  noDirectives: $$mapWrap(() => Recursive, \\"Query$noDirectives\\"),
};

type typeof$Query = typeof Query

const Recursive$$ = <F extends $$DeepArray<Recursive$>>(...frag: F): F[number][] => frag;

const _Recursive = {
  $: Recursive$$,
  id: \\"Recursive$id\\",
} as const;

export const Recursive: typeof _Recursive & {
  next: $$MapWrap<typeof$Recursive, \\"Recursive$next\\">,
} = {
  ..._Recursive,
  next: $$mapWrap(() => Recursive, \\"Recursive$next\\"),
};

type typeof$Recursive = typeof Recursive

const ShallowOnly$$ = <F extends $$DeepArray<ShallowOnly$>>(...frag: F): F[number][] => frag;

export const ShallowOnly = {
  $: ShallowOnly$$,
  a: \\"ShallowOnly$a\\",
  b: \\"ShallowOnly$b\\",
} as const;

type typeof$ShallowOnly = typeof ShallowOnly

export type $BinaryTree<F extends $$DeepArray<BinaryTree$>> = $$Result<\\"BinaryTree\\", $$UnwrapDeepArray<F>>
export type $DeepOnly<F extends $$DeepArray<DeepOnly$>> = $$Result<\\"DeepOnly\\", $$UnwrapDeepArray<F>>
export type $Query<F extends $$DeepArray<Query$>> = $$Result<\\"Query\\", $$UnwrapDeepArray<F>>
export type $Recursive<F extends $$DeepArray<Recursive$>> = $$Result<\\"Recursive\\", $$UnwrapDeepArray<F>>
export type $ShallowOnly<F extends $$DeepArray<ShallowOnly$>> = $$Result<\\"ShallowOnly\\", $$UnwrapDeepArray<F>>"
`;
